src/sandstorm/backend.c++

```diff
@@ -491,11 +520,32 @@ kj::Promise<void> BackendImpl::deletePackage(DeletePackageContext context) {
 // =======================================================================================
 
 kj::Promise<void> BackendImpl::backupGrain(BackupGrainContext context) {
+  kj::Maybe<Cgroup::FreezeHandle> freezeHandle;
+  kj::Maybe<kj::Own<kj::PromiseFulfiller<void>>> backupFulfiller;
+
   auto params = context.getParams();
 
+  auto grainId = params.getGrainId();
   auto path = kj::str("/var/sandstorm/backups/", params.getBackupId());
+
+  auto iter = supervisors.find(grainId);
+  if(iter == supervisors.end()) {
+    // Not running. Mark the grain as backing up, so nobody tries to boot it
+    // until we're done:
+    auto paf = kj::newPromiseAndFulfiller<void>();
+    supervisors.insert(std::make_pair(grainId, BackingUpGrain{paf.promise.fork()}));
+    backupFulfiller = kj::mv(paf.fulfiller);
+  } else {
+    // If cgroups is available, freeze the corresponding cgroup
+    // during the backup:
+    KJ_IF_MAYBE(cg, cgroup) {
+      auto grainCgroup = cg->getOrMakeChild(grainId);
```

Comment:
This `getOrMakeChild()` should always find an already-made cgroup, right? Maybe worth having a `getChild()` method that throws if it doesn't exist?

Reply:
Correct, done (also merged changes back in from master). Still need to do some more testing though.

Reply:
Correct, done (also merged some changes back in from master). Still need to do some more testing though.